java/ql/test/query-tests/security/CWE-502/GsonServlet.java

```diff
@@ -0,0 +1,77 @@
+import java.io.IOException;
+
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.typeadapters.RuntimeTypeAdapterFactory;
+
+import com.example.User;
+import com.thirdparty.Person;
+
+public class GsonServlet extends HttpServlet {
+
+    private static final long serialVersionUID = 1L;
+
+    @Override
+    // GOOD: concrete class type specified
+    public void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOException {
+        String json = req.getParameter("json");
+
+        Gson gson = new Gson();
+        Object obj = gson.fromJson(json, User.class);
+    }
+
+    @Override
+    // GOOD: concrete class type specified
+    public void doPut(HttpServletRequest req, HttpServletResponse resp) throws IOException {
+        String json = req.getParameter("json");
+
+        Gson gson = new Gson();
+        Object obj = gson.fromJson(json, Person.class);
+    }
+    
+    @Override
+    // BAD: allow class name to be controlled by remote source
+    public void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOException {
+        String json = req.getParameter("json");
+        String clazz = req.getParameter("class");
+
+        try {
+            Gson gson = new Gson();
+            Object obj = gson.fromJson(json, Class.forName(clazz));
+        } catch (ClassNotFoundException cne) {
+            throw new IOException(cne.getMessage());
+        }
+    }
+
+    @Override
+    // GOOD: specify allowed class types with trusted class
+    public void doHead(HttpServletRequest req, HttpServletResponse resp) throws IOException {
+        String json = req.getParameter("json");
+        String clazz = req.getParameter("class");
+
+        try {
+            RuntimeTypeAdapterFactory<User> runtimeTypeAdapterFactory = RuntimeTypeAdapterFactory
+            .of(User.class, "type");
+            Gson gson = new GsonBuilder().registerTypeAdapterFactory(runtimeTypeAdapterFactory).create();
+            Object obj = gson.fromJson(json, Class.forName(clazz));
```

Comment:
This is not actually safe, when `RuntimeTypeAdapterFactory` encounters a different type it will just let Gson try the next adapter / factory.

Reply:
Please see my comments above. I will make the proposed change.